tests/test_content_manifest.py

```diff
@@ -617,6 +617,75 @@ def test_to_json(mock_top_level_purl, app, package, subpath):
         mock_top_level_purl.assert_called_once_with(package, request, subpath=subpath)
 
 
+@pytest.mark.parametrize(
+    "package",
+    [
+        None,
+        {"name": "example.com/org/project", "type": "go-package", "version": "1.1.1"},
+        {"name": "gomod-some", "type": "gomod", "version": "1.0.0"},
+        {"name": "grc-ui", "type": "npm", "version": "1.0.0"},
+        {"name": "requests", "type": "pip", "version": "2.24.0"},
+        {"name": "grc-ui", "type": "yarn", "version": "1.0.0"},
+        {"name": "ruby-some", "type": "rubygems", "version": "1.0.0"},
+        {"name": "submodule-some", "type": "git-submodule", "version": "1.0.0"},
+    ],
+)
+@pytest.mark.parametrize("subpath", [None, "some/path"])
+@mock.patch("cachito.web.content_manifest.to_top_level_purl")
+def test_sbom_components_one_package(mock_top_level_purl, package, subpath):
+    request = Request()
+
+    if package and subpath:
+        package["path"] = subpath
+
+    packages = _load_packages_from_json([package]) if package else []
+
+    cm = ContentManifest(request, packages)
+    expected = {}
+
+    if package:
+        component = {
+            "type": "library",
+            "name": package["name"],
+            "version": package["version"],
+            "purl": mock_top_level_purl.return_value,
+        }
+        expected.setdefault(mock_top_level_purl.return_value, component)
+
+    assert expected == cm.sbom_components_dict()
+
+
+@pytest.mark.parametrize(
+    "packages_json",
+    [
+        [
+            {"name": "example.com/org/project", "type": "go-package", "version": "1.1.1"},
+            {"name": "grc-ui", "type": "npm", "version": "1.0.0"},
+            {"name": "requests", "type": "pip", "version": "2.24.0"},
+            {"name": "grc-ui2", "type": "yarn", "version": "1.0.0"},
+        ],
+    ],
+)
+@mock.patch("cachito.web.content_manifest.to_top_level_purl")
+def test_sbom_components_multiple_packages(mock_top_level_purl, packages_json):
+    request = Request()
+    packages = _load_packages_from_json(packages_json)
+
+    cm = ContentManifest(request, packages)
+    expected = {}
+
+    for package_json in packages_json:
+        component = {
+            "type": "library",
+            "name": package_json["name"],
+            "version": package_json["version"],
+            "purl": mock_top_level_purl.return_value,
+        }
+        expected.setdefault(mock_top_level_purl.return_value, component)
+
+    assert expected == cm.sbom_components_dict()
```

Comment:
Generating the expected data dynamically may not be a good idea. The SBOM should have 4 components, but has only 1. And you don't see that, because the expected data is generated.

Btw, de-duplicating by purl will be a problem. Top-level packages for pip, npm, yarn and ruby have the same purls (`pkg:github/org/repo@ref#subpath`)

Reply:
ok